File: src/content/docs/topics/cs/leetcode-150/binary-search/153-find-minimum-in-rotated-sorted-array.py
def find_min(nums: list) -> int:
    pass  # TODO: implement

def _run_tests():
    assert find_min([3, 4, 5, 1, 2]) == 1
    assert find_min([4, 5, 6, 7, 0, 1, 2]) == 0
    assert find_min([11, 13, 15, 17]) == 11
    assert find_min([1]) == 1
    assert find_min([2, 1]) == 1
    assert find_min([1, 2]) == 1
    # --- large-input timing ---
    import time as _t
    _arr = list(range(50000, 100000)) + list(range(0, 50000))
    _t0 = _t.perf_counter()
    find_min(_arr)
    _ms = (_t.perf_counter() - _t0) * 1000
    print(f'perf find_min rotated n=100000: {_ms:.1f}ms')
    print('all tests pass')

if __name__ == '__main__':
    _run_tests()

File: src/content/docs/topics/cs/coding-problems/binary-search/153-find-minimum-in-rotated-sorted-array.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

function findMin(nums: number[]): number {
    // TODO: implement
    return 0;
}

function _runTests(): void {
    assert(findMin([3, 4, 5, 1, 2]) === 1);
    assert(findMin([4, 5, 6, 7, 0, 1, 2]) === 0);
    assert(findMin([11, 13, 15, 17]) === 11);
    assert(findMin([1]) === 1);
    assert(findMin([2, 1]) === 1);
    assert(findMin([1, 2]) === 1);
    // perf
    const arr = [...Array.from({ length: 50000 }, (_, i) => i + 50000), ...Array.from({ length: 50000 }, (_, i) => i)];
    const t0 = performance.now();
    for (let i = 0; i < 1_000_000; i++) findMin(arr);
    console.log(`perf findMin rotated n=100000 x1000000: ${(performance.now() - t0).toFixed(1)}ms`);
    console.log('all tests pass');
}

_runTests();
